File: src/dere_plugins/dere_vault/scripts/link_analysis.py
#!/usr/bin/env python3
"""Analyze link density and knowledge graph health in permanent notes."""

from __future__ import annotations

import argparse
import re
import sys
from pathlib import Path

try:
    from .detect_vault import find_vault_root
except ImportError:
    from detect_vault import find_vault_root


def extract_wikilinks(content: str) -> set[str]:
    """Extract [[wikilinks]] from note content."""
    return set(re.findall(r"\[\[([^\]]+)\]\]", content))


def extract_frontmatter_tags(content: str) -> list[str]:
    """Extract tags from YAML frontmatter."""
    match = re.search(r"^---\n(.*?)\n---", content, re.DOTALL)
    if not match:
        return []

    frontmatter = match.group(1)
    tags = []
    in_tags = False

    for line in frontmatter.split("\n"):
        if line.startswith("tags:"):
            in_tags = True
            continue
        if in_tags:
            if line.startswith("  - "):
                tags.append(line.replace("  - ", "").strip())
            elif not line.startswith(" "):
                break

    return tags


def extract_title(content: str) -> str | None:
    """Extract title from frontmatter or first heading."""
    # Try frontmatter first
    match = re.search(r"^---\n(.*?)\n---", content, re.DOTALL)
    if match:
        title_match = re.search(r"^title:\s*[\"']?(.+?)[\"']?\s*$", match.group(1), re.MULTILINE)
        if title_match:
            return title_match.group(1)

    # Fall back to first heading
    heading_match = re.search(r"^#\s+(.+)$", content, re.MULTILINE)
    if heading_match:
        return heading_match.group(1)

    return None


def is_permanent_note(content: str) -> bool:
    """Check if note is a permanent note."""
    match = re.search(r"^---\n(.*?)\n---", content, re.DOTALL)
    if not match:
        return False

    frontmatter = match.group(1)
    return "type: permanent" in frontmatter


def analyze_vault(vault_path: Path) -> dict:
    """Analyze link structure of all permanent notes."""
    notes = {}
    note_titles = {}

    # First pass: collect all permanent notes and their titles
    for note_path in vault_path.rglob("*.md"):
        try:
            content = note_path.read_text(encoding="utf-8")
        except Exception:
            continue

        if not is_permanent_note(content):
            continue

        title = extract_title(content)
        if not title:
            continue

        note_key = note_path.stem
        note_titles[note_key] = title
        note_titles[title] = title  # Also map title to itself

        links = extract_wikilinks(content)
        tags = extract_frontmatter_tags(content)

        notes[note_key] = {
            "path": note_path.relative_to(vault_path),
            "title": title,
            "links_out": links,
            "links_in": set(),
            "tags": tags,
        }

    # Second pass: calculate incoming links
    for note_key, note_data in notes.items():
        for link in note_data["links_out"]:
            # Try to find the target note
            # Links could be to filename or to title
            target_key = link
            if link in note_titles:
                # Find the note with this title
                for k, data in notes.items():
                    if data["title"] == link:
                        target_key = k
                        break

            if target_key in notes:
                notes[target_key]["links_in"].add(note_key)

    return notes


def find_orphans(notes: dict, min_links: int = 3) -> list[dict]:
    """Find notes with fewer than min_links total links."""
    orphans = []

    for note_key, note_data in notes.items():
        total_links = len(note_data["links_out"]) + len(note_data["links_in"])
        if total_links < min_links:
            orphans.append({
                "title": note_data["title"],
                "path": note_data["path"],
                "links_out": len(note_data["links_out"]),
                "links_in": len(note_data["links_in"]),
                "total": total_links,
            })

    orphans.sort(key=lambda x: x["total"])
    return orphans


def suggest_connections(notes: dict, note_key: str, limit: int = 5) -> list[dict]:
    """Suggest potential connections for a note based on tag overlap."""
    if note_key not in notes:
        return []

    target_note = notes[note_key]
    target_tags = set(target_note["tags"])

    if not target_tags:
        return []

    suggestions = []

    for other_key, other_note in notes.items():
        if other_key == note_key:
            continue

        # Skip if already linked
        if other_key in target_note["links_out"] or other_key in target_note["links_in"]:
            continue

        other_tags = set(other_note["tags"])
        overlap = target_tags & other_tags

        if overlap:
            suggestions.append({
                "title": other_note["title"],
                "path": other_note["path"],
                "shared_tags": list(overlap),
                "overlap_count": len(overlap),
            })

    suggestions.sort(key=lambda x: x["overlap_count"], reverse=True)
    return suggestions[:limit]


def calculate_statistics(notes: dict) -> dict:
    """Calculate vault-wide link statistics."""
    if not notes:
        return {}

    link_counts = [len(n["links_out"]) + len(n["links_in"]) for n in notes.values()]
    out_counts = [len(n["links_out"]) for n in notes.values()]
    in_counts = [len(n["links_in"]) for n in notes.values()]

    return {
        "total_notes": len(notes),
        "avg_total_links": sum(link_counts) / len(link_counts),
        "avg_out_links": sum(out_counts) / len(out_counts),
        "avg_in_links": sum(in_counts) / len(in_counts),
        "max_links": max(link_counts),
        "min_links": min(link_counts),
    }


def main() -> int:
    """CLI entry point."""
    parser = argparse.ArgumentParser(description="Analyze knowledge graph link health")
    parser.add_argument("--vault-path", type=Path, help="Path to vault (auto-detected if not provided)")
    parser.add_argument("--orphans", action="store_true", help="Show orphaned notes (< 3 total links)")
    parser.add_argument("--suggest", type=str, help="Suggest connections for a note (by title or filename)")
    parser.add_argument("--stats", action="store_true", help="Show vault-wide statistics")
    parser.add_argument("--min-links", type=int, default=3, help="Minimum links for non-orphan status")

    args = parser.parse_args()

    # Find vault
    vault_path = args.vault_path
    if not vault_path:
        vault_path = find_vault_root()
        if not vault_path:
            print("Error: Could not find vault. Specify --vault-path", file=sys.stderr)
            return 1

    # Analyze
    print("Analyzing vault knowledge graph...\n")
    notes = analyze_vault(vault_path)

    if not notes:
        print("No permanent notes found in vault")
        return 0

    # Show statistics
    if args.stats or (not args.orphans and not args.suggest):
        stats = calculate_statistics(notes)
        print("Vault Statistics:")
        print(f"  Total permanent notes: {stats['total_notes']}")
        print(f"  Average total links: {stats['avg_total_links']:.1f}")
        print(f"  Average outgoing links: {stats['avg_out_links']:.1f}")
        print(f"  Average incoming links: {stats['avg_in_links']:.1f}")
        print(f"  Max links (any note): {stats['max_links']}")
        print(f"  Min links (any note): {stats['min_links']}")
        print()

    # Show orphans
    if args.orphans:
        orphans = find_orphans(notes, args.min_links)
        if orphans:
            print(f"Found {len(orphans)} orphaned note(s) with < {args.min_links} total links:\n")
            for i, orphan in enumerate(orphans, 1):
                print(f"{i}. {orphan['title']}")
                print(f"   Path: {orphan['path']}")
                print(f"   Links: {orphan['links_out']} out, {orphan['links_in']} in ({orphan['total']} total)")
                print()
        else:
            print(f"No orphaned notes found (all notes have >= {args.min_links} links)")

    # Suggest connections
    if args.suggest:
        # Find note by title or filename
        note_key = None
        for key, note_data in notes.items():
            if key == args.suggest or note_data["title"] == args.suggest:
                note_key = key
                break

        if not note_key:
            print(f"Error: Note not found: {args.suggest}", file=sys.stderr)
            return 1

        suggestions = suggest_connections(notes, note_key)
        if suggestions:
            print(f"Suggested connections for: {notes[note_key]['title']}\n")
            for i, suggestion in enumerate(suggestions, 1):
                print(f"{i}. {suggestion['title']}")
                print(f"   Path: {suggestion['path']}")
                print(f"   Shared tags: {', '.join(suggestion['shared_tags'])}")
                print()
        else:
            print(f"No suggestions found for: {notes[note_key]['title']}")
            print("(Note may already be well-connected or have no shared tags with other notes)")

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/dere_plugins/dere_vault/scripts/link_analysis.ts
import { readFile } from "node:fs/promises";
import { join, relative } from "node:path";

import { findVaultRoot } from "./detect_vault.js";

function extractWikiLinks(content: string): Set<string> {
  return new Set([...content.matchAll(/\[\[([^\]]+)\]\]/g)].map((match) => match[1]));
}

function extractFrontmatterTags(content: string): string[] {
  const match = content.match(/^---\n([\s\S]*?)\n---/);
  if (!match) {
    return [];
  }

  const frontmatter = match[1];
  const tags: string[] = [];
  let inTags = false;

  for (const line of frontmatter.split("\n")) {
    if (line.startsWith("tags:")) {
      inTags = true;
      continue;
    }
    if (inTags) {
      if (line.startsWith("  - ")) {
        tags.push(line.replace("  - ", "").trim());
      } else if (!line.startsWith(" ")) {
        break;
      }
    }
  }

  return tags;
}

function extractTitle(content: string): string | null {
  const frontmatterMatch = content.match(/^---\n([\s\S]*?)\n---/);
  if (frontmatterMatch) {
    const titleMatch = frontmatterMatch[1].match(/^title:\s*["']?(.+?)["']?\s*$/m);
    if (titleMatch?.[1]) {
      return titleMatch[1];
    }
  }

  const headingMatch = content.match(/^#\s+(.+)$/m);
  if (headingMatch?.[1]) {
    return headingMatch[1];
  }

  return null;
}

function isPermanentNote(content: string): boolean {
  const match = content.match(/^---\n([\s\S]*?)\n---/);
  if (!match) {
    return false;
  }
  return match[1].includes("type: permanent");
}

type NoteData = {
  path: string;
  title: string;
  links_out: Set<string>;
  links_in: Set<string>;
  tags: string[];
};

async function* walk(dir: string): AsyncGenerator<string> {
  const { readdir } = await import("node:fs/promises");
  let items;
  try {
    items = await readdir(dir, { withFileTypes: true });
  } catch {
    return;
  }
  for (const item of items) {
    const fullPath = join(dir, item.name);
    if (item.isDirectory()) {
      yield* walk(fullPath);
    } else if (item.isFile() && item.name.endsWith(".md")) {
      yield fullPath;
    }
  }
}

async function analyzeVault(vaultPath: string): Promise<Record<string, NoteData>> {
  const notes: Record<string, NoteData> = {};
  const noteTitles: Record<string, string> = {};

  for await (const notePath of walk(vaultPath)) {
    let content: string;
    try {
      content = await readFile(notePath, "utf-8");
    } catch {
      continue;
    }

    if (!isPermanentNote(content)) {
      continue;
    }

    const title = extractTitle(content);
    if (!title) {
      continue;
    }

    const noteKey = notePath.split("/").pop()?.replace(/\.md$/, "") ?? notePath;
    noteTitles[noteKey] = title;
    noteTitles[title] = title;

    const links = extractWikiLinks(content);
    const tags = extractFrontmatterTags(content);

    notes[noteKey] = {
      path: relative(vaultPath, notePath),
      title,
      links_out: links,
      links_in: new Set<string>(),
      tags,
    };
  }

  for (const [noteKey, noteData] of Object.entries(notes)) {
    for (const link of noteData.links_out) {
      let targetKey = link;
      if (noteTitles[link]) {
        for (const [key, data] of Object.entries(notes)) {
          if (data.title === link) {
            targetKey = key;
            break;
          }
        }
      }
      if (notes[targetKey]) {
        notes[targetKey].links_in.add(noteKey);
      }
    }
  }

  return notes;
}

function findOrphans(
  notes: Record<string, NoteData>,
  minLinks: number,
): Array<Record<string, unknown>> {
  const orphans: Array<Record<string, unknown>> = [];
  for (const noteData of Object.values(notes)) {
    const totalLinks = noteData.links_out.size + noteData.links_in.size;
    if (totalLinks < minLinks) {
      orphans.push({
        title: noteData.title,
        path: noteData.path,
        links_out: noteData.links_out.size,
        links_in: noteData.links_in.size,
        total: totalLinks,
      });
    }
  }
  orphans.sort((a, b) => (a.total as number) - (b.total as number));
  return orphans;
}

function suggestConnections(
  notes: Record<string, NoteData>,
  noteKey: string,
  limit: number,
): Array<Record<string, unknown>> {
  const targetNote = notes[noteKey];
  if (!targetNote) {
    return [];
  }

  const targetTags = new Set(targetNote.tags);
  if (targetTags.size === 0) {
    return [];
  }

  const suggestions: Array<Record<string, unknown>> = [];
  for (const [otherKey, otherNote] of Object.entries(notes)) {
    if (otherKey === noteKey) {
      continue;
    }
    if (targetNote.links_out.has(otherKey) || targetNote.links_in.has(otherKey)) {
      continue;
    }
    const otherTags = new Set(otherNote.tags);
    const overlap = [...targetTags].filter((tag) => otherTags.has(tag));
    if (overlap.length) {
      suggestions.push({
        title: otherNote.title,
        path: otherNote.path,
        shared_tags: overlap,
        overlap_count: overlap.length,
      });
    }
  }

  suggestions.sort((a, b) => (b.overlap_count as number) - (a.overlap_count as number));
  return suggestions.slice(0, limit);
}

function calculateStatistics(notes: Record<string, NoteData>): Record<string, number> {
  const noteList = Object.values(notes);
  if (!noteList.length) {
    return {};
  }

  const linkCounts = noteList.map((note) => note.links_out.size + note.links_in.size);
  const outCounts = noteList.map((note) => note.links_out.size);
  const inCounts = noteList.map((note) => note.links_in.size);

  return {
    total_notes: noteList.length,
    avg_total_links: linkCounts.reduce((sum, value) => sum + value, 0) / linkCounts.length,
    avg_out_links: outCounts.reduce((sum, value) => sum + value, 0) / outCounts.length,
    avg_in_links: inCounts.reduce((sum, value) => sum + value, 0) / inCounts.length,
    max_links: Math.max(...linkCounts),
    min_links: Math.min(...linkCounts),
  };
}

async function main(): Promise<void> {
  const args = process.argv.slice(2);
  let vaultPath: string | null = null;
  let showOrphans = false;
  let suggestTitle: string | null = null;
  let showStats = false;
  let minLinks = 3;

  for (let i = 0; i < args.length; i += 1) {
    const arg = args[i];
    if (arg === "--vault-path") {
      vaultPath = args[i + 1] ?? null;
      i += 1;
    } else if (arg === "--orphans") {
      showOrphans = true;
    } else if (arg === "--suggest") {
      suggestTitle = args[i + 1] ?? null;
      i += 1;
    } else if (arg === "--stats") {
      showStats = true;
    } else if (arg === "--min-links") {
      minLinks = Number.parseInt(args[i + 1] ?? "3", 10);
      i += 1;
    }
  }

  const resolvedVault = vaultPath ?? findVaultRoot();
  if (!resolvedVault) {
    console.error("Error: Could not find vault. Specify --vault-path");
    process.exit(1);
  }

  console.log("Analyzing vault knowledge graph...\n");
  const notes = await analyzeVault(resolvedVault);

  if (!Object.keys(notes).length) {
    console.log("No permanent notes found in vault");
    process.exit(0);
  }

  if (showStats || (!showOrphans && !suggestTitle)) {
    const stats = calculateStatistics(notes);
    console.log("Vault Statistics:");
    console.log(`  Total permanent notes: ${stats.total_notes}`);
    console.log(`  Average total links: ${stats.avg_total_links.toFixed(1)}`);
    console.log(`  Average outgoing links: ${stats.avg_out_links.toFixed(1)}`);
    console.log(`  Average incoming links: ${stats.avg_in_links.toFixed(1)}`);
    console.log(`  Max links (any note): ${stats.max_links}`);
    console.log(`  Min links (any note): ${stats.min_links}`);
    console.log("");
  }

  if (showOrphans) {
    const orphans = findOrphans(notes, minLinks);
    if (orphans.length) {
      console.log(`Found ${orphans.length} orphaned note(s) with < ${minLinks} total links:\n`);
      orphans.forEach((orphan, index) => {
        console.log(`${index + 1}. ${orphan.title}`);
        console.log(`   Path: ${orphan.path}`);
        console.log(
          `   Links: ${orphan.links_out} out, ${orphan.links_in} in (${orphan.total} total)`,
        );
        console.log("");
      });
    } else {
      console.log(`No orphaned notes found (all notes have >= ${minLinks} links)`);
    }
  }

  if (suggestTitle) {
    let noteKey: string | null = null;
    for (const [key, noteData] of Object.entries(notes)) {
      if (key === suggestTitle || noteData.title === suggestTitle) {
        noteKey = key;
        break;
      }
    }

    if (!noteKey) {
      console.error(`Error: Note not found: ${suggestTitle}`);
      process.exit(1);
    }

    const suggestions = suggestConnections(notes, noteKey, 5);
    if (suggestions.length) {
      console.log(`Suggested connections for: ${notes[noteKey].title}\n`);
      suggestions.forEach((suggestion, index) => {
        console.log(`${index + 1}. ${suggestion.title}`);
        console.log(`   Path: ${suggestion.path}`);
        console.log(`   Shared tags: ${(suggestion.shared_tags as string[]).join(", ")}`);
        console.log("");
      });
    } else {
      console.log(`No suggestions found for: ${notes[noteKey].title}`);
      console.log("(Note may already be well-connected or have no shared tags with other notes)");
    }
  }
}

if (import.meta.main) {
  void main();
}
